app/models/unit.rb

```diff
@@ -92,13 +92,49 @@ def role_for(user)
   has_many :convenors, -> { joins(:role).where('roles.name = :role', role: 'Convenor') }, class_name: 'UnitRole'
   has_many :staff, ->     { joins(:role).where('roles.name = :role_convenor or roles.name = :role_tutor', role_convenor: 'Convenor', role_tutor: 'Tutor') }, class_name: 'UnitRole'
 
+  # Unit has a teaching period
+  belongs_to :teaching_period
+  validates :start_date, presence: true
+  validates :end_date, presence: true
+
+  validate :validate_end_date_after_start_date
+  validate :ensure_teaching_period_dates_match, if: :has_teaching_period?
+
   scope :current,               -> { current_for_date(Time.zone.now) }
   scope :current_for_date,      ->(date) { where('start_date <= ? AND end_date >= ?', date, date) }
   scope :not_current,           -> { not_current_for_date(Time.zone.now) }
   scope :not_current_for_date,  ->(date) { where('start_date > ? OR end_date < ?', date, date) }
   scope :set_active,            -> { where('active = ?', true) }
   scope :set_inactive,          -> { where('active = ?', false) }
 
+  def teaching_period_id=(teaching_period_id)
+    if teaching_period_id.present?
+      tp = TeachingPeriod.find(teaching_period_id)
+      write_attribute(:start_date, tp.start_date)
+      write_attribute(:end_date, tp.end_date)
+    end
+    super(teaching_period_id)
+  end
+
+  def has_teaching_period?
+    self.teaching_period.present?
+  end
+
+  def ensure_teaching_period_dates_match
```

Comment:
If this fails... what can the user do?

Reply:
If this fails, the API will throw an error.
Practically, this will never happen, because whenever there is a teaching period, I am manually overriding the value of start and end date to make it same.

Reply:
This should probably just be in a unit test then...

Reply:
May be better to override the setters for the dates and raise errors there if they are set when there is a teaching period... though this complicates the setting of these when we do set the teaching period :)